src/mds/Server.cc

```diff
@@ -1391,59 +1401,73 @@ void Server::reconnect_tick()
     return;
   }
 
-  utime_t reconnect_end = reconnect_start;
-  reconnect_end += g_conf()->mds_reconnect_timeout;
-  if (ceph_clock_now() >= reconnect_end &&
-      !client_reconnect_gather.empty()) {
-    dout(10) << "reconnect timed out" << dendl;
+  if (client_reconnect_gather.empty())
+    return;
 
-    // If we're doing blacklist evictions, use this to wait for them before
-    // proceeding to reconnect_gather_finish
-    MDSGatherBuilder gather(g_ceph_context);
+  auto now = clock::now();
+  auto elapse = std::chrono::duration<double>(now - reconnect_start).count();
+  if (elapse < g_conf()->mds_reconnect_timeout)
+    return;
 
-    for (set<client_t>::iterator p = client_reconnect_gather.begin();
-	 p != client_reconnect_gather.end();
-	 ++p) {
-      Session *session = mds->sessionmap.get_session(entity_name_t::CLIENT(p->v));
-      ceph_assert(session);
-
-      // Keep sessions that have specified timeout. These sessions will prevent
-      // mds from going to active. MDS goes to active after they all have been
-      // killed or reclaimed.
-      if (session->info.client_metadata.find("timeout") !=
-	  session->info.client_metadata.end()) {
-	dout(1) << "reconnect keeps " << session->info.inst << ", need to be reclaimed" << dendl;
-	client_reclaim_gather.insert(session->get_client());
-	continue;
-      }
+  vector<Session*> remaining_sessions;
+  remaining_sessions.reserve(client_reconnect_gather.size());
+  for (auto c : client_reconnect_gather) {
+    Session *session = mds->sessionmap.get_session(entity_name_t::CLIENT(c.v));
+    ceph_assert(session);
+    remaining_sessions.push_back(session);
+    // client re-sends cap flush messages before the reconnect message
+    if (session->last_seen > reconnect_last_seen)
+      reconnect_last_seen = session->last_seen;
+  }
 
-      dout(1) << "reconnect gives up on " << session->info.inst << dendl;
+  elapse = std::chrono::duration<double>(now - reconnect_last_seen).count();
+  if (elapse < g_conf()->mds_reconnect_timeout / 2)
+    return;
 
-      mds->clog->warn() << "evicting unresponsive client " << *session
-                        << ", after waiting " << g_conf()->mds_reconnect_timeout
-                        << " seconds during MDS startup";
+  dout(10) << "reconnect timed out" << dendl;
```

Comment:
Let's make this message more descriptive: some client sessions have not reconnected in time and will be evicted.